### PR TITLE
Rocketchip bug fixes + zcu102 softcore

### PR DESCRIPTION
**Pull request by chrisguikema**
This PR does two things:

1. Differentiates the base Rocketchip platform with a version generated for the ZCU102 FPGA fabric
2. Fixes bugs we encountered running on the softcore implementation

We tested this with the VMM created by @yyshen

### CODE REVIEW

File: src/arch/riscv/common_riscv.lds
Comment (axel-h) [inline src/arch/riscv/common_riscv.lds:82]: Can you add a more verbose explanation here, why this padding is needed on your platforms. I have removed this, because it was completely unclear, why this is needed. Bringing it back is fine, but this time it must have a proper explanation, what is going on on this platform.

File: src/arch/riscv/kernel/boot.c
Comment (axel-h) [inline src/arch/riscv/kernel/boot.c:242]: Can you be more specific, what is the problem then? For the user mode mapping, we will have to use full pages anyway. Maybe we should run up for all platforms?

Comment (chrisguikema) [reply src/arch/riscv/kernel/boot.c:242]: I didn't go super in-depth here to figure out _why_ this was happening, but I can tell you the behavior we were seeing.

We ran yanyan's VMM just fine with an older codebase. When rebasing onto this branch, we ran into a couple issues. One was the 8 byte padding, the other was that user-space generated an "virtual address already mapped" error. By reverting a few old commits, things would work properly, but that's not a good solution for larger commits in my opinion.

I was able to get past the virtual address error by increasing the static allocator mempool size, but I'd immediately get a RAM fault when trying to run the VM. 

So I looked at what was happening in the commits we needed to revert, and one of them removed the rounding for the device tree. Like you mentioned, userspace needs to map in full pages anyways, so if a portion of a page is reserved, the entire page should be. My best guess is that there was an overlap somewhere there shouldn't have been, but like I said, I didn't grab GDB and look at every memory map, I just found the solution.

File: src/arch/riscv/common_riscv.lds
Comment (chrisguikema) [reply src/arch/riscv/common_riscv.lds:82]: I just tested it again, and its the same behavior I described with the DTB round up. The rootserver gets an "Virtual address already mapped" exception from the kernel.

**Comment by chrisguikema** [comment]
By the way, we did all of our testing of the riscv_he_v0.6 branch, but these changes should also apply to `master` if discussion warrants it

File: src/object/interrupt.c
Comment (Ivan-Velickovic) [inline src/object/interrupt.c:138]: 1. Do you know if this apply to all versions of QEMU?
2. This should also apply to `CONFIG_PLAT_QEMU_RISCV_VIRT`

Comment (chrisguikema) [reply src/object/interrupt.c:138]: @Furao do you know the answer to this? We can update the ifdef easy enough.

File: src/arch/riscv/machine/hardware.c
Comment (axel-h) [inline src/arch/riscv/machine/hardware.c:171]: Please remove this, this function is already defined in `include/arch/riscv/arch/sbi.h`, see https://github.com/seL4/seL4/blob/master/include/arch/riscv/arch/sbi.h#L101-L104

File: src/object/interrupt.c
Comment (axel-h) [reply src/object/interrupt.c:138]: It's better to use explicit platform checks, so it is clear where this applies. Please check for `CONFIG_PLAT_QEMU_RISCV_VIRT` for now, as this is the official QEMU platform. 
There are other platforms (HiFive, PolarFire) where a QEMU simulation also exists. They may have the same problem, but here we should consider the real hardware as the default target. An additional kernel config define should be added to explicitly enable the fix for this specific QEMU bug.

File: src/arch/riscv/machine/hardware.c
Comment (chrisguikema) [reply src/arch/riscv/machine/hardware.c:171]: Done.

File: src/object/interrupt.c
Comment (chrisguikema) [reply src/object/interrupt.c:138]: I agree that we should do an explicit check for QEMU_RISCV_VIRT, but that platform doesn't exist in `riscv_he_v0.6`, hence the check for `spike`.

@axel-h @Ivan-Velickovic do you think this PR should be rebased onto master?

**Comment by chrisguikema** [comment]
@yyshen I think a rebase may have conflicted our commit `91f35b9b789cebb72d7d75eea8c9e3ff5d8b6d69` with your commit `bd660ffb71b756791e0b26ebfbbcbe558dae5944`. I'll do some testing to see what changes actually need to be made for the system to run on real hardware.

**Comment by axel-h** [comment]
@chrisguikema: I have a PR pending to update OpenSBI, see https://github.com/seL4/seL4/pull/383
What version are you using?

**Comment by chrisguikema** [comment]
> What version are you using?

We're using this branch of an OpenSBI fork: https://github.com/bao-project/opensbi/tree/bao/rocket

File: src/object/interrupt.c
Comment (Ivan-Velickovic) [reply src/object/interrupt.c:138]: Ah, my bad. I forgot that this branch wouldn't have support for virt yet. Yes, I think it's a good idea to rebase the branch on master.

Comment (chrisguikema) [reply src/object/interrupt.c:138]: Just to clarify, you're saying I should rebase `rocketchip_bug_fixes_zcu102` onto master, not `riscv_he_0.6`?

Comment (axel-h) [reply src/object/interrupt.c:138]: It seems @Ivan-Velickovic ist trying to rebase `riscv_he_0.6`, so you can rebease on it then.

Comment (chrisguikema) [reply src/object/interrupt.c:138]: I rebased this branch onto `master`

File: src/object/interrupt.c
Comment (axel-h) [inline src/object/interrupt.c:140]:
```suggestion
     * platforms, the claim can wait and be done here.
```

**Comment by chrisguikema** [comment]
Once this PR is approved, I can link the resources we've used to run on the ZCU102 softcore rocketchip. It would probably be best to update the docs page, right?

File: src/arch/riscv/common_riscv.lds
Comment (chrisguikema) [reply src/arch/riscv/common_riscv.lds:92]: is it possible to wrap this change around the CONFIG_PLAT_ROCKETCHIP_ZCU102 ifdef? That way its only used for the platform its needed.

Comment (axel-h) [reply src/arch/riscv/common_riscv.lds:92]: In https://github.com/seL4/seL4/blob/master/CMakeLists.txt#L657-L669 the linker script run through the preprocessor, so an `#ifdef` could work.

File: src/object/interrupt.c
Comment (chrisguikema) [reply src/object/interrupt.c]: Done.

File: src/arch/riscv/machine/hardware.c
Comment (axel-h) [inline src/arch/riscv/machine/hardware.c:106]: Can you extend the comment to mention the counterpart in `invokeIRQHandler_AckIRQ()`, like the comment there mentions this. Then we have the connection, in case QEMU's behavior changes in future version.

Comment (chrisguikema) [reply src/arch/riscv/machine/hardware.c:106]: Done.

File: src/arch/riscv/common_riscv.lds
Comment (axel-h) [reply src/arch/riscv/common_riscv.lds:98]: I've created https://github.com/seL4/seL4/issues/972 as a follow-up task to investigate this. So it's not necessarily a merge blocker to have this conditional code for now.

File: src/arch/riscv/kernel/boot.c
Comment (axel-h) [reply src/arch/riscv/kernel/boot.c:241]: I've created https://github.com/seL4/seL4/issues/972 as a follow-up task to investigate this. So it's not necessarily a merge blocker to have this conditional code for now.

File: src/arch/riscv/common_riscv.lds
Comment (chrisguikema) [reply src/arch/riscv/common_riscv.lds:98]: Thanks, Axel!

File: src/arch/riscv/kernel/boot.c
Comment (chrisguikema) [reply src/arch/riscv/kernel/boot.c:241]: Thanks, Axel!

**Comment by chrisguikema** [comment]
Just wanted to ping this PR to make sure it didn't get stale. @Ivan-Velickovic have you had a chance to take another look?

**Comment by axel-h** [comment]
This looks mostly ok from my side. I just wonder where `ROCKETCHIP_ZCU102` actually comes from. Is this running against a released version from https://github.com/chipsalliance/rocket-chip that we could mention in a comment in `rocketchip/config.cmake` or `overlay-rocketchip-zcu102.dts` That would at least give some pointers until somebody add some details how to build and run this are at https://docs.sel4.systems/Hardware/rocketchip.html

File: src/plat/rocketchip/config.cmake
Comment (axel-h) [inline src/plat/rocketchip/config.cmake:37]: Mabye add the comment here that needs the OpenSBI fork from https://github.com/bao-project/opensbi/tree/bao/rocket then, as it seem not mainlines yet.

Comment (chrisguikema) [reply src/plat/rocketchip/config.cmake:37]: Good suggestion; I'll add a link.

**Comment by chrisguikema** [comment]
> This looks mostly ok from my side. I just wonder where `ROCKETCHIP_ZCU102` actually comes from. 

I didn't generate the bitstream myself, but one of my co-workers did. I think @Furao or @FTGStudio may have more insight. Basically, we added the h-extensions to the mainline rocketchip and generated a bitstream compatible with the ZCU102. My plan is to release the raw bitstream with instructions such that anyone with a ZCU102 can grab the bitstream and get it up-and-running quickly.

Because there were differences with the hardware resources (memory location, interrupt controller, etc), I created the "new platform" to keep the existing rocketchip support from changing.

**Comment by Furao** [comment]
Chris is pretty close. We are using the `hyp` branch of this fork of rocket-chip which has hyp extensions applied: https://github.com/josecm/rocket-chip. This branch is at some point after v1.6. We also apply some patches to the mainline chipyard to add the ZCU102 as a supported platform: https://github.com/ucb-bar/chipyard/tree/64632c8aeef5386b4840e324be4429c0ceaac635 

We basically replicated the steps here: https://github.com/bao-project/bao-demos

**Comment by chrisguikema** [comment]
My plan is to update the docs site with a PR once this is approved that points to a pre-built image, as well as the steps required to replicate the image.

**Comment by axel-h** [comment]
Thanks @Furao and @chrisguikema. Can you add all these comment into the code/script, so we have this documented at least in a place where somebody picking up this patch will find it easily.
I consider the QEMU fix is something to merge in an case. It might be worth bringing this into a separate PR to merge this quickly (assuming there is an agreement to merge this). CI testing should also show that this works fine with the RISC-V hardware that is there.
For the rest, seems it's stuck where some PRs that add new platform get stuck. Needs some input from @lsf37 now on how to proceed.

**Comment by chrisguikema** [comment]
I've added comments in the code such that the discussion points made here aren't lost. I've also created a rocketchip-zcu102.md file which I can make a PR for in the docs repo. This document overviews how we generated the bitstream such that others can reproduce it.

**Comment by chrisguikema** [comment]
I created a docs PR here: https://github.com/seL4/docs/pull/175

File: src/object/interrupt.c
Comment (kent-mcleod) [inline src/object/interrupt.c:144]: Can this extra return statement please be removed as it's changing the AST of verified code and causing the preprocess tests to fail.

**Comment by kent-mcleod** [comment]
> Differentiates the base Rocketchip platform with a version generated for the ZCU102 FPGA fabric

The description for the base platform is also for a ZCU FPGA fabric (https://docs.sel4.systems/Hardware/rocketchip.html)? Why is this a different platform? Is it just a different version of the bitstream with particular extensions? If this is the case, then maybe the config names could be changed to better reflect this?

**Comment by chrisguikema** [comment]
> The description for the base platform is also for a ZCU FPGA fabric

The ZC706 is a zynq7000 development kit. The ZCU102 is a zynqmp development kit. I'm not an FPGA guy, but I assume there are differences in the FPGA fabric which make the two incompatible

File: src/plat/rocketchip/config.cmake
Comment (axel-h) [inline src/plat/rocketchip/config.cmake:40]: So, is it zcu102 or zcu104 now?

```suggestion
        # The rocket-fpga-zcu102 platform can be found at the following git repo:
```

File: src/plat/rocketchip/overlay-rocketchip-zcu102.dts
Comment (axel-h) [inline src/plat/rocketchip/overlay-rocketchip-zcu102.dts:35]: Could you add a linebreak in the last line?

File: src/plat/rocketchip/config.cmake
Comment (chrisguikema) [reply src/plat/rocketchip/config.cmake:40]: The repo we forked has it as `rocket-zcu-104`. 

https://github.com/dornerworks/opensbi/tree/bao/rocket/platform/rocket-fpga-zcu104

Comment (chrisguikema) [reply src/plat/rocketchip/config.cmake:40]: But it works without issues on the ZCU102 hardware

File: src/plat/rocketchip/overlay-rocketchip-zcu102.dts
Comment (chrisguikema) [reply src/plat/rocketchip/overlay-rocketchip-zcu102.dts]: Done.

File: src/plat/rocketchip/config.cmake
Comment (axel-h) [reply src/plat/rocketchip/config.cmake:40]: Ok, maybe add this also to the text, so this is clear.

**Review by kent-mcleod — APPROVED** [review]

**Pull request merged**
